instrumentation/http/src/main/java/brave/http/HttpRequestParser.java

```diff
@@ -0,0 +1,86 @@
+ * the License.
+ */
+package brave.http;
+
+import brave.SpanCustomizer;
+import brave.propagation.ExtraFieldPropagation;
+import brave.propagation.TraceContext;
+
+/**
+ * Use this to control the request data recorded for an {@link TraceContext#sampledLocal() sampled
+ * HTTP client or server span}.
+ *
+ * <p>Here's an example that changes the span name and records the HTTP url instead of the path.
+ * <pre>{@code
+ * httpTracing = httpTracing.toBuilder()
+ *   .clientRequestParser((request, context, span) -> {
+ *     span.name(spanName(adapter, request)); // default span name
+ *     span.tag("http.url", request.url()); // the whole url, not just the path
+ *   }).build();
+ * }</pre>
+ *
+ * <h3>Relationship to {@link HttpParser}</h3>
+ * This handles the request side of what {@link HttpParser} formerly did. This uses the {@link
+ * HttpRequest} type directly, avoiding the deprecated {@link HttpAdapter} type. By constraining
+ * this to only request data, it is implementable as a lambda or method reference. This also passes
+ * the trace context as a parameter, lowering parsing overhead vs relying on the current span.
+ *
+ * <h3>The {@link TraceContext} parameter</h3>
+ * The {@link TraceContext} parameter provides for advanced data handling, such as {@link
+ * ExtraFieldPropagation#get(TraceContext, String) extra fields}. This is explicitly passed as you
+ * cannot rely on {@link brave.Tracer#currentSpan()} being the same value.
+ *
+ * @see HttpResponseParser
+ */
+public interface HttpRequestParser {
+  HttpRequestParser DEFAULT = new HttpRequestParser.Default();
+
+  /**
+   * Implement to choose what data from the http request are parsed into the span representing it.
+   *
+   * @see Default
+   */
+  void parse(HttpRequest request, TraceContext context, SpanCustomizer span);
+
+  /**
+   * The default data policy sets the span name to the HTTP method and adds the "http.method" and
+   * "http.path" tags.
+   */
+  // Eventhough the default span name is the method, we have no way of knowing that a user hasn't
+  // overwritten the name to something else. If that occurs during response parsing, it is too late
+  // to go back and get the http method. Adding http method by default ensures span naming doesn't
+  // prevent basic HTTP info from being visible. A cost of this is another tag, but it is small with
+  // very limited cardinality. Moreover, users who care strictly about size can override this.
+  class Default implements HttpRequestParser {
+    /**
+     * This sets the span name to the HTTP method and adds the "http.method" and "http.path" tags.
+     *
+     * <p>If you only want to change the span name, subclass and override {@link
+     * #spanName(HttpRequest, TraceContext)}.
+     */
+    @Override public void parse(HttpRequest req, TraceContext context, SpanCustomizer span) {
+      String name = spanName(req, context);
+      if (name != null) span.name(name);
```

Comment:
For example here, the fact that method is allowed to be null makes this single line of code a lot more complex (what are the implications of not setting name, for example)

Reply:
spanName can be overridden, so it can return null anyway (because it allowed this in the past). For example, you can intentionally return null to only set the name at response time. Anyway, I get your point about method though.

Reply:
https://github.com/openzipkin/brave/issues/1089